nfs/src/rpc_task.cpp

```diff
@@ -484,163 +484,147 @@ void access_callback(
 
-            args.file = client->get_nfs_inode_from_ino(ino)->get_fh();
-            args.offset = mb->offset + bc->pvt;
-            args.count  = mb->length - bc->pvt;
-            args.stable = FILE_SYNC;
-            args.data.data_len = mb->length - bc->pvt;
-            args.data.data_val = (char *) mb->buffer + bc->pvt;
+            AZLogDebug("[{}] Partial write: [{}, {}) of [{}, {})", ctx->get_ino(),
+                       ctx->get_offset(), ctx->get_length(),
+                       ctx->get_offset(), res->WRITE3res_u.resok.count);
 
             /*
-             * Create new flush task to write remaining buffer.
-             * New task help us to account stats accurately.
-             * This flush task will be continuing the job of writing to the
-             * same bc, so set bc and ctx.
+             * Free the bytes_chunk which are written fully.
              */
-            struct rpc_task *flush_task =
-                client->get_rpc_task_helper()->alloc_rpc_task(FUSE_FLUSH);
-            flush_task->init_flush(nullptr /* fuse_req */, ino);
-            flush_task->rpc_api->bc = bc;
-            flush_task->rpc_api->pvt = (void *) ctx;
-
-            // Update the ctx with new task.
-            ctx->set_task(flush_task);
+            auto it = bc_vec.begin();
+            while (it != bc_vec.cend()) {
+                uint64_t off = it->offset + it->pvt;
+                uint64_t len = it->length - it->pvt;
 
-            // TODO: Make this AZLogDebug after some time.
-            AZLogInfo("[{}] Partial write: [{}, {}) of [{}, {})", ctx->get_ino(),
-                       mb->offset + bc->pvt, mb->offset + mb->length,
-                       mb->offset, mb->offset + mb->length);
+                // This bc is written fully in this write.
+                if ((off + len) <= end_off) {
+                    /*
+                     * Get the membuf and do the sanity check before setting uptodate flag.
+                     */
+                    struct membuf* mb = it->get_membuf();
+                    assert(mb != nullptr);
+                    assert(mb->is_inuse() && mb->is_locked());
+                    assert(mb->is_flushing() && mb->is_dirty() && mb->is_uptodate());
+
+                    mb->clear_dirty();
+                    mb->clear_flushing();
+                    mb->clear_locked();
+                    mb->clear_inuse();
+                    rpc_iov->iov++;
+                    rpc_iov->iovcnt--;
+                    rpc_iov->offset += len;
+                    rpc_iov->length -= len;
+
+                    // We don't want to cache the data after the write completes.
+                    inode->filecache_handle->release(it->offset, it->length);
+
+                    // erase the entry from bc_vec.
+                    it = bc_vec.erase(it);
+                } else {
+                    assert(end_off >= off);
+                    it->pvt = (end_off - off);
```

Comment:
bc->pvt is the running offset into the bc, so this is WRONG. It should actually be (end_off - it->offset), but then you need another variable for following calculations. Partial writes will result in data corruption due to this.